### ISSUE TITLE
tf.Print related wired bug

### ISSUE THREAD

**Issue by NaxAlpha** [tensorflow/tensorflow#20773]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 10 (latest)
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: r1.9 (v1.9.0-0-g25c197e023)
- **Python version**: Python 3.6.5 |Anaconda, Inc.| [MSC v.1900 64 bit (AMD64)] on win32
- **Bazel version (if compiling from source)**: N/A
- **GCC/Compiler version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: CUDA 9.0, cuDNN 64_7
- **GPU model and memory**: GTX 1060-6GB(Laptop)
- **Exact command to reproduce**: N/A

### Describe the problem
While developing a prime number generator using `tensorflow-gpu`, I came across a wired bug. Given the source code, if you use `tf.Print` result is correct and run-time is less than without using `tf.Print`. 

### Source code / logs
``` python
# check first 'n' numbers for primity using tensorflow

import warnings
with warnings.catch_warnings():
    warnings.filterwarnings("ignore",category=FutureWarning)
    import h5py
import os
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3' 
import tensorflow as tf
import numpy as np
import time

def prime_tf_graf(max_count,dump):
	ttyp = tf.int32
	data = tf.range(2, max_count, dtype=ttyp)
	if dump:
		data = tf.Print(data, [data], 'Data Loaded!')
	X, Y = tf.meshgrid(data, data); 
	temp = Y%X
	mask = tf.cast(tf.equal(temp, Y), dtype=ttyp)
	temp = temp - Y * mask
	temp = tf.cast(tf.not_equal(temp, 0), dtype=ttyp)
	sumr = tf.reduce_sum(temp, axis=1)
	nums = data - 2
	rato = tf.cast(sumr/nums, ttyp)
	indx =  tf.cast(tf.not_equal(rato, 0), ttyp)
	shap =  tf.reshape(tf.where(indx), [-1]) + 2
	return shap

def prime_tf(max_count,dump):
	with tf.device('/gpu:0'):
		graf = prime_tf_graf(max_count, dump)
	sess = tf.Session(config=tf.ConfigProto(allow_soft_placement=True))
	sess.run(tf.global_variables_initializer())
	prim = sess.run(graf)
	return prim

def bench(task):
	now = time.time()
	res = task()
	return res, time.time() - now

p, t = bench(lambda: prime_tf(10000, True))
print('With Dump: found %d primes in %f time'%(len(p), t))
p, t = bench(lambda: prime_tf(10000, False))
print('Without Dump: found %d primes in %f time'%(len(p), t))
```
and here is output on my system:

```
PS E:\Research\TFLearn> python .\bug.py
Data Loaded![2 3 4...]
With Dump: found 1229 primes in 12.599950 time
Without Dump: found 1085 primes in 25.162465 time
```

**Comment by poxvoculi** [comment]
What happens if you switch ttyp from tf.int32 to tf.int64?  I.e. don't try to use int32 on GPU.

**Comment by NaxAlpha** [comment]
Still similar response with performance is worse this time:
```
Data Loaded![2 3 4...]
With Dump: found 1229 primes in 21.802515 time
Without Dump: found 1085 primes in 109.847477 time
```

**Comment by poxvoculi** [comment]
I consistently get a div-by-0 error at:

	rato = tf.cast(sumr/nums, ttyp)

**Comment by NaxAlpha** [comment]
That may be possible but on my system, it doesn't give any error! 
Using this instead: `rato = tf.cast(tf.equal(sumr, nums), ttyp)` gives correct result but performance is still worse without `tf.Print`:
```
Data Loaded![2 3 4...]
With Dump: found 1229 primes in 13.345328 time
Without Dump: found 1229 primes in 22.766805 time
```
I think it may be because tf.Print somehow forces tensor to load quickly but that's just my assumption.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @poxvoculi, @zheng-xq: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @poxvoculi, @zheng-xq: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @poxvoculi, @zheng-xq: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by poxvoculi** [comment]
@aaroey Can you take a look at this.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @poxvoculi, @aaroey: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @poxvoculi, @aaroey: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @poxvoculi, @aaroey: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @poxvoculi, @aaroey: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by mohantym** [comment]
Hi @NaxAlpha ! I am getting the same value now in 2.8 version . Attaching [gist ](https://colab.sandbox.google.com/gist/mohantym/8c8c11a533d5fe08f4558576bd6631b5/untitled349.ipynb)for reference. Thanks!
